README.md
<div align="center">
<p align="center">
<img src="https://raw.githubusercontent.com/jorgemxm/mock-rest-server/master/.gitlab/MockRestServer.svg" width="256" height="256" alt="Mock REST Server"/>
</p>

<p align="center">
<h3>Mock REST Server is a lightweight (in memory) REST server for unit test.</h3>
</p>

<p align="center">
<a href="https://snyk.io/test/npm/mock-rest-server"><img src="https://snyk.io/test/npm/mock-rest-server/badge.svg" alt="Known Vulnerabilities"></a>
<a href="https://standardjs.com"><img src="https://img.shields.io/badge/code_style-standard-brightgreen.svg" alt="JavaScript Standard Style"></a>
<a href="https://www.npmjs.com/package/mock-rest-server"><img src="https://img.shields.io/npm/v/mock-rest-server.svg" alt="NPM version"></a>
<a href="https://gitlab.com/GuilleW/mock-rest-server/-/blob/master/LICENSE"><img src="https://img.shields.io/npm/l/mock-rest-server.svg" alt="License"></a>
</p>
<br/>
</div>

# Mock REST Server

No configuration needed: Just start it, and test your code!

Mock REST Server comes with these features 🚀 :
- [x] Support GET, POST, PUT, PATCH, DELETE and OPTIONS HTTP methods
- [x] Accept and respond with JSON
- [x] Support real HTTP status code response from REST API
- [x] Return HTTP error status codes via api version
- [x] Fake latency
- [x] Filtering, sorting and pagination
- [x] Auto populate tool to fill database, don't waste your time for a REST server!

# Changes

All [details of changes](https://github.com/jorgemxm/mock-rest-server/blob/master/CHANGELOG.md) to this project will be documented in this file.

# Installation & Usage

```sh
npm i -g https://github.com/jorgemxm/mock-rest-server
node_modules/.bin/mock-server
  or
node_modules/.bin/mserver
```

| Param name | Description | Type | Default |
| ----------- | ----------- | ---- | ------- |
| `-p3000 --port=3000` | (Optional) Change server port | `Number` | `3000` |
| `-s     --silent` | (Optional) Hide server output | `Boolean` | `false` |
| `-l500  --latency` | (Optional) wait (ms) before response | `Number` | `0` |

## Unit test

For more details, look at the [full example](https://github.com/jorgemxm/mock-rest-server/blob/master/test/MockRestServer.spec.js) from test file.

Assuming you're using a module-compatible system (like [webpack](https://webpack.js.org/)),
start MockRestServer on top of your unit test file:

```js
import MockRestServer from 'mock-rest-server'

describe('MockRestServer', function () {
  it('Start server', async () => {
    const server = await MockRestServer.start(3000, true, 0)
    server.populate('articles', 30, {
      title: String,
      body: String,
      userId: Number,
      created: Date,
      private: Boolean
    })

    // ...
    // ...
    // ...

    // don't forget to stop!
    // If you work with `--watch` param that reload your unit test, don't forget to stop server at the end of your tests.
    // It'll try to launch server again when refresh, and port will already in use.
    await server.stop()
  })
})
```

MockRestServer come with (optional) awesome feature that fill database with random typed data: `populate(collection, length, schema)`

| Param name | Description | Type |
| ----------- | ----------- | ---- |
| `collection` | Collection name to create/populate | `String` |
| `length` | Number of resource to create in collection | `Number` |
| `schema` | Object send to create a resource. Values are replaced by random value based on JS type `String`, `Number`, `Date` or `Boolean`. | `Object` |

# Example

Start MockRestServer, open a new shell and run some curl on api `/v1/`:
```sh
curl -X POST -d '{"title":"Awesome movie!","body":"Some content."}' http://localhost:3000/v1/movies
# {"title":"Awesome news!","body":"Some content.","id":1}
```

Now, get your movies with:
```sh
curl http://localhost:3000/v1/movies
# [{"title":"Awesome news!","body":"Some content.","id":1},{"title":"Awesome news!","body":"Some content.","id":2}]
curl http://localhost:3000/v1/movies/1
# {"title":"Awesome news!","body":"Some content.","id":1}
```

Use api `/v[xxx]/` to mock HTTP status codes (403, 404, 500...) from server response:
```sh
curl http://localhost:3000/v401/books
curl http://localhost:3000/v200/superheroes
```

# Contribution

Any help or feedback are really welcome, **no matter how great or small**!

Please make sure to read the [Contributing Guide](https://github.com/jorgemxm/mock-rest-server/blob/master/CONTRIBUTING.md) before making a pull request.

# License

[GPL-3.0](https://gitlab.com/GuilleW/mock-rest-server/-/blob/master/LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0]
### Added
- Add option to mock latency for each query

## [1.1.5] - 2020-09-18
### Changed
- Improve unit test: support for HTTP OPTIONS method
- Improve unit test: if the JSON in body request is not valid JSON, respond with a Bad Request 400 Error.
### Fixed
- Throws a Bad Request 400 Error if the JSON in body request to parse is not valid JSON.

## [1.1.4] - 2020-09-10
### Fixed
- Add support for HTTP OPTIONS method so CORS is happy

## [1.1.3] - 2020-09-08
### Fixed
- Bad import package name in the README documentation

## [1.1.2] - 2020-09-06
### Changed
- Use [From Changelog to NPM](https://gitlab.com/GuilleW/from-changelog-to-npm) to publish on NPM
### Fixed
- watch.yml is missing

## [1.1.1] - 2020-08-07
### Fixed
- Gitlab CI not publishing on npmjs.com

## [1.1.0] - 2020-08-07
### Added
- Use CI for Release and Publish on npmjs.com

## [1.0.1] - 2020-07-23
### Changed
- Improve header of README.md with logo
### Fixed
- Text in logo not vectorized

## [1.0.0] - 2020-07-22
### Added
- Mock REST Server published on [gitlab](https://gitlab.com/GuilleW/mock-rest-server) and [npmjs](https://www.npmjs.com/package/mock-rest-server).


test/MockRestServer.spec.js
/* global describe it */

const expect = require('chai').expect
const MockRestServer = require('../src/MockRestServer')
const fetch = require('node-fetch')
const url = 'http://localhost:3000'
var server

const dataset = [
  { title: 'foo', body: 'bar', userId: 1 },
  { title: 'lorem', body: 'ipsum', userId: 1 },
  { title: 'my', body: 'awesome text', userId: 2 }
]

describe('MockRestServer should', function () {
  it('Start server', async () => {
    await fetch(url)
      .then(response => {
        expect(response.errno).to.equal('ECONNREFUSED')
      })
      .catch((response) => {
        expect(response.errno).to.equal('ECONNREFUSED')
      })
    server = await MockRestServer.start(3000, true)
    await fetch(url)
      .then(response => {
        expect(response.status).to.equal(400)
      })
      .catch(err => console.log(err))
    await fetch(url, {
      method: 'OPTIONS'
    })
      .then(response => {
        expect(response.status).to.equal(200)
      })
      .catch(err => console.log(err))
  })
  it('Request POST     (Create a resource)', async () => {
    await dataset.forEach(async (item, i) => {
      const id = i + 1
      await fetch(`${url}/v1/articles`, {
        method: 'POST',
        body: JSON.stringify(item),
        headers: {
          'Content-type': 'application/json; charset=UTF-8'
        }
      })
        .then(response => {
          expect(response.status).to.equal(201)
          return response.json()
        })
        .then(json => {
          expect(json.id).to.equal(id)
          expect(json.title).to.equal(item.title)
          expect(json.body).to.equal(item.body)
          expect(json.userId).to.equal(item.userId)
        })
        .catch(err => console.log(err))
    })
  })
  it('Request GET      (List all resources)', async () => {
    await fetch(`${url}/v1/articles`)
      .then(response => {
        expect(response.status).to.equal(200)
        return response.json()
      })
      .then(json => {
        dataset.forEach((item, i) => {
          item.id = i + 1
          expect(json[i].id).to.equal(item.id)
          expect(json[i].title).to.equal(item.title)
          expect(json[i].body).to.equal(item.body)
          expect(json[i].userId).to.equal(item.userId)
        })
      })
      .catch(err => console.log(err))
  })
  it('Request GET      (Get a resource)', async () => {
    await dataset.forEach(async (item, i) => {
      const id = i + 1
      await fetch(`${url}/v1/articles/${id}`)
        .then(response => {
          expect(response.status).to.equal(200)
          return response.json()
        })
        .then(json => {
          expect(json.id).to.equal(id)
          expect(json.title).to.equal(item.title)
          expect(json.body).to.equal(item.body)
          expect(json.userId).to.equal(item.userId)
        })
        .catch(err => console.log(err))
    })
  })
  it('Request PUT      (Create a resource)', async () => {
    await fetch(`${url}/v1/users/2`, {
      method: 'PUT',
      body: JSON.stringify({ user: 'Admin', password: '****' }),
      headers: {
        'Content-type': 'application/json; charset=UTF-8'
      }
    })
      .then(response => {
        expect(response.status).to.equal(200)
        return response.json()
      })
      .then(json => {
        expect(json.id).to.equal(2)
        expect(json.user).to.equal('Admin')
        expect(json.password).to.equal('****')
      })
      .catch(err => console.log(err))
  })
  it('Request PUT      (Replace a resource)', async () => {
    await dataset.forEach(async (item, i) => {
      const id = i + 1
      item.id = id
      delete item.title
      item.body += ' ' + item.id
      await fetch(`${url}/v1/articles/${id}`, {
        method: 'PUT',
        body: JSON.stringify(item),
        headers: {
          'Content-type': 'application/json; charset=UTF-8'
        }
      })
        .then(response => {
          expect(response.status).to.equal(200)
          return response.json()
        })
        .then(json => {
          expect(json.id).to.equal(item.id)
          expect(json.title).to.equal(undefined)
          expect(json.body).to.equal(item.body)
          expect(json.userId).to.equal(item.userId)
        })
        .catch(err => console.log(err))
    })
  })
  it('Request PATCH    (Create a resource)', async () => {
    await fetch(`${url}/v1/world/2`, {
      method: 'PATCH',
      body: JSON.stringify({ password: '*' }),
      headers: {
        'Content-type': 'application/json; charset=UTF-8'
      }
    })
      .then(response => {
        expect(response.status).to.equal(200)
        return response.json()
      })
      .then(json => {
        expect(json.id).to.equal(2)
        expect(json.password).to.equal('*')
      })
      .catch(err => console.log(err))
  })
  it('Request PATCH    (Update a resource)', async () => {
    await dataset.forEach(async (item, i) => {
      const id = i + 1
      item.id = id
      item.body += '!'
      await fetch(`${url}/v1/articles/${id}`, {
        method: 'PATCH',
        body: JSON.stringify({ body: item.body }),
        headers: {
          'Content-type': 'application/json; charset=UTF-8'
        }
      })
        .then(response => {
          expect(response.status).to.equal(200)
          return response.json()
        })
        .then(json => {
          expect(json.id).to.equal(item.id)
          expect(json.title).to.equal(undefined)
          expect(json.body).to.equal(item.body)
          expect(json.userId).to.equal(item.userId)
        })
        .catch(err => console.log(err))
    })
  })
  it('Request with filtering', async () => {
    await fetch(`${url}/v1/articles?userId=1`)
      .then(response => {
        expect(response.status).to.equal(200)
        return response.json()
      })
      .then(json => {
        var len = 0
        json.forEach((item, i) => {
          expect(item.userId).to.equal(1)
          len++
        })
        expect(len).to.equal(2)
      })
      .catch(err => console.log(err))
  })
  it('Request with sorting', async () => {
    await fetch(`${url}/v1/articles?sort=+body`)
      .then(response => {
        expect(response.status).to.equal(200)
        return response.json()
      })
      .then(json => {
        expect(json[0].id).to.equal(3)
        expect(json[1].id).to.equal(1)
        expect(json[2].id).to.equal(2)
      })
      .catch(err => console.log(err))
    await fetch(`${url}/v1/articles?sort=-body`)
      .then(response => {
        expect(response.status).to.equal(200)
        return response.json()
      })
      .then(json => {
        expect(json[0].id).to.equal(2)
        expect(json[1].id).to.equal(1)
        expect(json[2].id).to.equal(3)
      })
      .catch(err => console.log(err))
  })
  it('Populate database', async () => {
    server.populate('articles', 30, {
      title: String,
      body: String,
      userId: Number,
      created: Date,
      private: Boolean
    })
    await fetch(`${url}/v1/articles`)
      .then(response => {
        expect(response.status).to.equal(200)
        return response.json()
      })
      .then(json => {
        expect(json[3].title).to.be.a('string')
        expect(json[3].body).to.be.a('string')
        expect(json[3].userId).to.be.a('number')
        expect(json[3].created).to.be.a('string')
        expect(json[3].private).to.be.a('boolean')
        expect(json.length).to.equal(30 + dataset.length)
      })
      .catch(err => console.log(err))
  })
  it('Request with pagination', async () => {
    await fetch(`${url}/v1/articles?page=2`)
      .then(response => {
        expect(response.status).to.equal(200)
        return response.json()
      })
      .then(json => {
        for (let i = 0; i < 10; i++) {
          expect(json[i].id).to.equal(i + 11)
        }
      })
      .catch(err => console.log(err))
  })
  it('Request DELETE   (Delete a resource)', async () => {
    await fetch(`${url}/v1/articles/1`, {
      method: 'DELETE'
    })
      .then(response => {
        expect(response.status).to.equal(204)
      })
      .catch(err => console.log(err))
    await fetch(`${url}/v1/articles/1`)
      .then(response => {
        expect(response.status).to.equal(404)
      })
      .catch(err => console.log(err))
  })
  it('Request DELETE   (Delete collection)', async () => {
    await fetch(`${url}/v1/articles`, {
      method: 'DELETE'
    })
      .then(response => {
        expect(response.status).to.equal(204)
      })
      .catch(err => console.log(err))
    await fetch(`${url}/v1/articles`)
      .then(response => {
        expect(response.status).to.equal(404)
        return response.json()
      })
      .catch(err => console.log(err))
  })
  it('Respond with 500 (Internal Server Error)', async () => {
    await fetch(`${url}/v500/articles`)
      .then(response => {
        expect(response.status).to.equal(500)
      })
      .catch(err => console.log(err))
  })
  it('Respond with 400 (Broken JSON → Bad Request Error)', async () => {
    await fetch(`${url}/v1/articles`, {
      method: 'POST',
      body: '{a:1, $b:2}'
    })
      .then(response => {
        expect(response.status).to.equal(400)
      })
      .catch(err => console.log(err))
  })
  it('Stop server', async () => {
    await fetch(url)
      .then(response => {
        expect(response.status).to.equal(400)
      })
      .catch(err => console.log(err))
    await server.stop()
    await fetch(url)
      .then(response => {
        expect(response.errno).to.equal('ECONNREFUSED')
      })
      .catch((response) => {
        expect(response.errno).to.equal('ECONNREFUSED')
      })
  })
})

describe('MockRestServer should run in verbose mode', function () {
  it('Verbose mode work', async () => {
    server = await MockRestServer.start()
    await dataset.forEach(async (item, i) => {
      await fetch(`${url}/v1/articles`, {
        method: 'POST',
        body: JSON.stringify(item),
        headers: {
          'Content-type': 'application/json; charset=UTF-8'
        }
      })
        .then(response => {
          expect(response.status).to.equal(201)
        })
        .catch(err => console.log(err))
      await server.stop()
    })
  })
})


.gitlab/ci/Release.js
const fs = require('fs')
const { execSync } = require('child_process')
const request = require('./request')

const content = fs.readFileSync('CHANGELOG.md', { encoding: 'utf8', flag: 'r' })
const changelog = content.split('\n## ')[2] || ''
const version = (changelog.match(/\[(.+?)\]/) || [])[1] || ''

// Update package.json version
console.log(execSync(`git config user.email "${process.env.GITLAB_USER_EMAIL}"`).toString())
console.log(execSync(`git config user.name "${process.env.GITLAB_USER_NAME}"`).toString())
console.log(execSync(`npm version "${version}" -m "Release v${version}"`).toString())
console.log(execSync(`git push https://gitlab-ci-token:${process.env.GIT_PUSH_TOKEN}@gitlab.com/${process.env.CI_PROJECT_PATH}.git HEAD:${process.env.CI_COMMIT_REF_NAME}`).toString())

// Create new Tag/Release
;(async () => {
  await request(`https://gitlab.com/api/v4/projects/${process.env.CI_PROJECT_ID}/releases`, {
    name: `Release v${version}`,
    ref: process.env.CI_COMMIT_SHA,
    tag_name: `v${version}`,
    description: `## ${changelog}`.replace(/(["'’`])/g, '\\$1')
  })
})()


src/MockRestServer.js
const http = require('http')
const Database = require('./Database')
const Randomizer = require('./Randomizer')

class MockRestServer {
  constructor (port, silent, latency) {
    this.port = port || 3000
    this.silent = silent || false
    this.latency = latency || 0
    this.db = new Database()
    this.trace('Mock Rest Server: \n')
    this.trace('Open a new shell and run some curl on api /v1/:')
    this.trace(`  → curl -X POST -d '{"title":"Awesome movie!","body":"Some content."}' http://localhost:${this.port}/v1/movies\n`)
    this.trace('Now, get your movies with:')
    this.trace(`  → curl http://localhost:${this.port}/v1/movies`)
    this.trace(`  → curl http://localhost:${this.port}/v1/movies/1\n`)
    this.trace('Use api /v[xxx]/ to mock HTTP status codes (403,404,500...) from server response:')
    this.trace(`  → curl http://localhost:${this.port}/v401/books`)
    this.trace(`  → curl http://localhost:${this.port}/v200/superheroes\n`)


    this.trace(`Mock Rest Server is ready and listen at port ${this.port}:`)
    this.server = this.startServer(this.port)
  }

  static start (port, silent, latency) {
    return new MockRestServer(port, silent, latency)
  }

  startServer (port) {
    const self = this
    return http.createServer((req, res) => {
      self.request(req, res)
    }).listen(port)
  }

  stop () {
    this.server.close()
  }

  populate (collection, length, schema) {
    var resource = () => {
      var json = {}
      for (const key in schema) {
        json[key] = Randomizer.get(schema[key])
      }
      return json
    }

    for (let i = 0; i < length; i++) {
      this.db.create(collection, resource())
    }
  }

  trace (str) {
    if (!this.silent) console.log(str)
  }

  paramParser (req) {
    const [query, param] = req.url.split('?')
    const [, version, collection, id] = query.split('/')
    const field = ((param && param.replace(/(page|sort)=[^&]+/g, '').match(/(\w+)=/)) || [])[1]
    const filter = ((param && param.replace(/(page|sort)=[^&]+/g, '').match(/=([^&]+)/)) || [])[1]
    const sort = ((param && param.match(/sort=([+-][a-z]+)/)) || [])[1]
    const page = ((param && param.match(/page=([0-9]+)/)) || [])[1]
    return {
      version: parseInt(version.replace('v', '')),
      collection,
      id: parseInt(id),
      field,
      filter,
      sort,
      page
    }
  }

  responseStatusCodes (req, json) {
    const jsonIsEmpty = Object.keys(json).length === 0

    // 2xx Success
    //
    // 200 OK
    if (req.method === 'GET' && (!jsonIsEmpty || Array.isArray(json))) {
      return 200
    }
    if (req.method === 'PUT' && !jsonIsEmpty) {
      return 200
    }
    if (req.method === 'PATCH' && !jsonIsEmpty) {
      return 200
    }

    // 201 Created
    if (req.method === 'POST') {
      return 201
    }

    // 204 No Content
    if (req.method === 'DELETE') {
      return 204
    }

    // 4xx Client errors
    //
    // 404 Not Found
    return 404
  }

  route (req, res, json) {
    const param = this.paramParser(req)

    if (req.method === 'OPTIONS') {
      return this.response(req, res, 200, {})
    }

    if (isNaN(param.version)) {
      return this.response(req, res, 400, {})
    }

    if (param.version !== 1) {
      return this.response(req, res, param.version, {})
    }

    if (req.method === 'POST') {
      json = this.db.create(param.collection, json)
      return this.response(req, res, this.responseStatusCodes(req, json), json)
    }

    if (req.method === 'GET') {
      json = this.db.read(param.collection, param.id)

      // Filtering
      if (param.field && param.filter) {
        var filtered = []
        json.forEach((item, i) => {
          if (item[param.field] && RegExp(item[param.field]).test(param.filter)) {
            filtered.push(item)
          }
          json = filtered
        })
      }

      // Sorting
      if (param.sort) {
        const asc = param.sort[0] === '+'
        const field = param.sort.slice(1)
        if (asc) {
          json.sort((a, b) => a[field].localeCompare(b[field]))
        } else {
          json.sort((a, b) => b[field].localeCompare(a[field]))
        }
      }

      // Pagination
      if (param.page) {
        const page = Math.max(1, parseInt(param.page) || 1)
        json = json.slice((page - 1) * 10, page * 10)
      }

      return this.response(req, res, this.responseStatusCodes(req, json), json)
    }

    if (req.method === 'PUT') {
      json.id = param.id
      json = this.db.replace(param.collection, json)
      return this.response(req, res, this.responseStatusCodes(req, json), json)
    }

    if (req.method === 'PATCH') {
      json.id = param.id
      json = this.db.update(param.collection, json)
      return this.response(req, res, this.responseStatusCodes(req, json), json)
    }

    if (req.method === 'DELETE') {
      json = this.db.delete(param.collection, param.id)
      return this.response(req, res, this.responseStatusCodes(req, json), json)
    }
  }

  request (req, res) {
    const self = this
    var data = ''
    req
      .on('data', d => {
        data += d
      })
      .on('end', () => {
        var json = {}
        try {
          json = data !== '' ? JSON.parse(data) : {}
        } catch (e) {
          return self.response(req, res, 400, {})
        }
        self.route(req, res, json)
      })
  }

  response (req, res, statusCode, json) {
    const headers = {
      'Access-Control-Allow-Credentials': true,
      'Access-Control-Allow-Headers': '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, PATCH, DELETE, OPTIONS',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Max-Age': 3600,
      'Content-Type': 'application/json; charset=UTF-8'
    }
    json = JSON.stringify(json)

    this.trace(`  - ${req.method.padEnd(6, ' ')} ${req.url} → [${statusCode}] ${json}`)

    if (statusCode < 100) {
      return res.end('Invalid Status Code')
    }

    res.writeHead(statusCode, headers)
    setTimeout(() => res.end(json), this.latency)
  }
}

module.exports = MockRestServer
